Add torch.sparse.as_sparse_gradcheck decorator of gradcheck that allows gradcheck input function to receive and return sparse tensors

pearu opened the pull request:
Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107903
* __->__ #107150
* #107777
* #107638



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

pearu committed:
Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"




[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

albanD commented on torch/sparse/__init__.py:
```diff
@@ -525,3 +525,178 @@ def func_wrapper(*args, **kwargs):
+                    else:
+                        # Materialize unspecified elements with zero values
+                        full_nnz = obj.shape[:obj.sparse_dim()].numel()
+                        full_values = torch.zeros((full_nnz, *values.shape[1:]), dtype=values.dtype, device=values.device)
```
Can't we use `obj.size()` to get the full size without this trick?
Or is the point to collapse all sparse dims similar to doing `.view(-1, *values.shape[1:])` ?

I'm not sure why you do this since it makes the indexing below hard.

Note that for both this one and the one below, can we use our add() function to get this in an easier way?

```python
full_val = torch.zeros(obj.size())
full_val = full_val + obj
```

pearu replied:
We can actually use other tricks to materialize COO tensor zeros:
```python
full_obj = torch.ones(obj.shape, dtype=obj.dtype, device=obj.device).to_sparse(layout=torch.sparse_coo, dense_dim=obj.dense_dim())
full_obj.values().sub_(1)
full_obj += obj
```
or
```python
full_nnz = obj.shape[:obj.sparse_dim()].numel()
full_indices = torch.ones(obj.shape[:obj.sparse_dim()], device=indices.device, dtype=torch.int8).nonzero().to(dtype=indices.dtype).T
full_values = torch.zeros((full_nnz, *obj.shape[obj.sparse_dim():]), device=values.device, dtype=values.dtype)
full_obj = torch.sparse_coo_tensor(full_indices, full_values, obj.shape)._coalesced_(True)
full_obj += obj
```
Performance-wise these appear to be equivalent, the former is a bit simpler, imho, so I'll pick this as a resolution.

pearu replied:
Unfortunately, the same approach does not work for CSR tensors because adding CSR tensors with batch dimensions is broken, see https://github.com/pytorch/pytorch/issues/107478.

pearu replied:
Applied the hints for COO tensors. For CSR tensors no changes are made because there exists bugs in adding CSR tensors.

albanD reviewed (commented):
Some high level question:
- Why is this on top of the outputs PR? This new function can replace it.
-

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented on torch/sparse/__init__.py:
```diff
@@ -525,3 +525,178 @@ def func_wrapper(*args, **kwargs):
     return gradcheck_with_sparse_outputs
+
+
+def enable_sparse_support(gradcheck):
```
`as_sparse_gradcheck` SGTM.

pearu replied:
Applied.

pearu commented on torch/sparse/__init__.py:
```diff
@@ -525,3 +525,178 @@ def func_wrapper(*args, **kwargs):
+            strided_outputs = tuple(outputs) if isinstance(outputs, (list, tuple)) else (outputs,)
+            strided_outputs = tuple((torch.Tensor.to_dense(o, masked_grad=masked_grad)
+                                     if o.requires_grad and o.layout in sparse_layouts else o)
+                                    for o in strided_outputs)
```
Not ok. The outputs ought to be anything std gradcheck supports.  tI'll fix it.

pearu replied:
Fixed.

pearu commented on torch/sparse/__init__.py:
```diff
@@ -525,3 +525,178 @@ def func_wrapper(*args, **kwargs):
+            if d['layout'] is torch.sparse_coo:
+                # TODO: After fixing gh-107097, replace `.coalesce()`
+                # method call with `._coalesced_(True)`.
+                return torch.sparse_coo_tensor(d['indices'], values, size=d['shape']).coalesce()
```
Ideally, it should be `._coalesced_(d['original'].is_coalesced())` but using this is blocked by https://github.com/pytorch/pytorch/issues/107097.

Otherwise, using `coalesce()` is not required here indeed.

albanD replied:
Ok!

pearu replied:
coalesce is removed, looking forward to https://github.com/pytorch/pytorch/issues/107097 fix.

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented on torch/sparse/__init__.py:
```diff
@@ -525,3 +525,178 @@ def func_wrapper(*args, **kwargs):
+            outputs = func(*restored_args, **kwargs)
+
+            strided_outputs = tuple(outputs) if isinstance(outputs, (list, tuple)) else (outputs,)
+            strided_outputs = tuple((torch.Tensor.to_dense(o, masked_grad=masked_grad)
```
Applied.

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented:
> * Why is this on top of the outputs PR? This new function can replace it.

The outputs PR is now removed from the stack.

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

albanD commented on docs/source/sparse.rst:
```diff
@@ -1332,6 +1332,15 @@ To manage checking sparse tensor invariants, see:
     sparse.check_sparse_tensor_invariants
 
+To enable sparse output support for :func:`~torch.autograd.gradcheck`
+function, decorate the function with
```
I don't think that this renders the way you want: https://docs-preview.pytorch.org/pytorch/pytorch/107150/sparse.html#other-functions

pearu replied:
Fixed, please review.

albanD reviewed (approved):
I would expect we could make the input creation simpler, but maybe not :(

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

albanD reviewed (approved):
Update looks good!

pearu committed:
Update on "Add torch.sparse.enable_sparse_support decorator of gradcheck that allows gradcheck input function to recieve and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.as_sparse_gradcheck decorator of gradcheck that allows gradcheck input function to receive and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.as_sparse_gradcheck decorator of gradcheck that allows gradcheck input function to receive and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge

pearu committed:
Update on "Add torch.sparse.as_sparse_gradcheck decorator of gradcheck that allows gradcheck input function to receive and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge

pearu committed:
Update on "Add torch.sparse.as_sparse_gradcheck decorator of gradcheck that allows gradcheck input function to receive and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.as_sparse_gradcheck decorator of gradcheck that allows gradcheck input function to receive and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu committed:
Update on "Add torch.sparse.as_sparse_gradcheck decorator of gradcheck that allows gradcheck input function to receive and return sparse tensors"

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.





cc alexsamardzic nikitaved cpuhrsch amjames bhosmer ezyang albanD zou3519 gqchen soulitzer Lezcano Varal7

[ghstack-poisoned]

pearu commented:
@pytorchbot merge
